mesonbuild/interpreter.py

```diff
@@ -1695,7 +1695,7 @@ def get_cross_property_method(self, args, kwargs):
 class Interpreter(InterpreterBase):
 
     def __init__(self, build, backend=None, subproject='', subdir='', subproject_dir='subprojects',
-                 modules = None, default_project_options=[]):
+                 modules = None, default_project_options={}):
```

Comment:
The default value should be `None` and then add`if default_project_options is None: default_project_options = {}` at the start. This is a known Python gotcha. (Even if it were OK to do here it is better to do the `None` thing consistently everywhere so people don't spend brain cycles ensuring the usage is ok in the future.)

This also changes the semantics a bit. The old code made a copy of the data that was passed in, this one uses a reference. If Python had `const` we could ensure that the option list is not altered

Reply:
I don't understand what's wrong in having {} in default value, but None if fine too.

About the copy, indeed I dropped the [:] initially because I didn't know how to do the same with dict and I it's not strictly needed anymore, but since then I learned dict has a copy() method (thanks @nirbheek) so I added that and rebased on master.

Reply:
> I don't understand what's wrong in having {} in default value, but None if fine too.

http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments